doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+* Handling batch requests.
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `WEBHOOK` source, and add a new endpoint to
+`environmentd`'s HTTP server `/api/webhook/:database/:schema/:name`. When a user creates this new
+kind of source, we'll persist the new `Source` in the catalog, use the `storage-client` to create a
+new collection.
+
+When we receive a event we'll use the `mz_adapter::Client` that the HTTP server has to send the raw
+`headers` and `body` to the coordinator, pack the data into a row, and append it to the necessary
+collection using a new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `WEBHOOK`, for which you can
+specify a `BODY FORMAT`, this denotes how we decode the body of a request. To start we'll support
+three formats, `BYTES`, `JSON`, or `TEXT`. By default we will not include the headers of the
+request, they can optionally be added by specifying `INCLUDE HEADERS`.
+
+```
+CREATE SOURCE <name> FROM WEBHOOK
+  BODY FORMAT [BYTES | JSON | TEXT]
+  (INCLUDE HEADERS)?
+;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type                        | optional?                                      |
+|---------|-----------------------------|------------------------------------------------|
+| body    | `bytea`, `jsonb`, or `text` | No                                             |
+| headers | `map[text -> text]`         | Yes, present if `INCLUDE HEADERS` is specified |
+
+We'll add a new endpoint to the existing [`base_router`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/environmentd/src/http.rs#L747)
+in `environmentd` with the path: `/api/webhook/:database/:schema/:name`. This follows the existing
+pattern of our other two endpoints, whose paths also start with `/api`. Users can then send events
+to their source, using its fully qualified name.
+
+> **_Note_**: The existing Web Sockets API is at `/api/experimental/ws`, we could move the webhooks
+API under the path `/api/experimental/webhook`, if we think there will be significant breaking
+changes in the future.
+
+We'll also add a new API `mz_adapter::Client::append_webhook` which is very similar to the existing
+[`insert_rows`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/adapter/src/client.rs#L392-L397)
+on the `SessionClient`. This new API will send the following to the `Coordinator` via a new `Command`:
```

Comment:
We don't want to start adding new coordinator messages that will increase the number of data rows the coordinator sees. This API should instead send the db, schema, name only, and the coordinator returns a `Arc<Catalog>` and the controller channel. This would also resolve the optional headers question below. The webhook caller can then do steps 4-7 below itself.

Reply:
This is great! I changed the architecture I currently have to match this and this also results in much better organized code!